ORC might not be working

Hi,

ORC reader might not be working (Trino 406). TSV/CSV works as expected.

```sql
SELECT * FROM
storage.orc."https://raw.githubusercontent.com/snowlift/trino-storage/master/src/test/resources/example-data/apache-lz4.orc";
```

```
2023-01-30T20:15:24.593Z	ERROR	Query-20230130_201524_00000_hnezj-162	org.ebyhr.trino.storage.StorageClient	Failed to get table: orc.https://raw.githubusercontent.com/snowlift/trino-storage/master/src/test/resources/example-data/apache-lz4.orc
java.lang.RuntimeException: java.io.FileNotFoundException: /tmp/trino-storage-6308204577453222815.orc (No such file or directory)
	at org.ebyhr.trino.storage.operator.OrcPlugin.getReader(OrcPlugin.java:119)
	at org.ebyhr.trino.storage.operator.OrcPlugin.getFields(OrcPlugin.java:55)
	at org.ebyhr.trino.storage.StorageClient.getTable(StorageClient.java:73)
	at org.ebyhr.trino.storage.StorageMetadata.getTableHandle(StorageMetadata.java:71)
	at org.ebyhr.trino.storage.StorageMetadata.getTableHandle(StorageMetadata.java:39)
	at io.trino.spi.connector.ConnectorMetadata.getTableHandle(ConnectorMetadata.java:121)
	at io.trino.metadata.MetadataManager.lambda$getTableHandle$4(MetadataManager.java:281)
	at java.base/java.util.Optional.flatMap(Optional.java:289)
	at io.trino.metadata.MetadataManager.getTableHandle(MetadataManager.java:275)
	at io.trino.metadata.MetadataManager.getRedirectionAwareTableHandle(MetadataManager.java:1516)
	at io.trino.metadata.MetadataManager.getRedirectionAwareTableHandle(MetadataManager.java:1508)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.getTableHandle(StatementAnalyzer.java:5360)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitTable(StatementAnalyzer.java:2149)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitTable(StatementAnalyzer.java:476)
	at io.trino.sql.tree.Table.accept(Table.java:60)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:493)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.analyzeFrom(StatementAnalyzer.java:4416)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:2878)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:476)
	at io.trino.sql.tree.QuerySpecification.accept(QuerySpecification.java:155)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:493)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:501)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:1449)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:476)
	at io.trino.sql.tree.Query.accept(Query.java:107)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:493)
	at io.trino.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:455)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:79)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:71)
	at io.trino.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:267)
	at io.trino.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:204)
	at io.trino.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:856)
	at io.trino.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:143)
	at io.trino.$gen.Trino_406____20230130_201510_2.call(Unknown Source)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:74)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.FileNotFoundException: /tmp/trino-storage-6308204577453222815.orc (No such file or directory)
	at java.base/java.io.RandomAccessFile.open0(Native Method)
	at java.base/java.io.RandomAccessFile.open(RandomAccessFile.java:344)
	at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:259)
	at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:213)
	at io.trino.orc.FileOrcDataSource.<init>(FileOrcDataSource.java:30)
	at org.ebyhr.trino.storage.operator.OrcPlugin.getReader(OrcPlugin.java:116)
	... 42 more
```

--- a/src/main/java/org/ebyhr/trino/storage/operator/OrcPlugin.java
+++ b/src/main/java/org/ebyhr/trino/storage/operator/OrcPlugin.java
@@ -31,7 +31,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.StandardCopyOption;
-import java.util.LinkedList;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 import java.util.function.Function;
@@ -51,62 +51,66 @@ public class OrcPlugin
     @Override
     public List<StorageColumn> getFields(String path, Function<String, InputStream> streamProvider)
     {
-        path = getLocalPath(path, streamProvider);
-        OrcReader reader = getReader(new File(path));
-        ColumnMetadata<OrcType> types = reader.getFooter().getTypes();
-        return reader.getRootColumn().getNestedColumns().stream()
-                .map(orcColumn -> new StorageColumn(
-                        orcColumn.getColumnName(),
-                        fromOrcType(types.get(orcColumn.getColumnId()), types)))
-                .collect(Collectors.toList());
-    }
-
-    @Override
-    public Iterable<Page> getPagesIterator(String path, Function<String, InputStream> streamProvider)
-    {
-        path = getLocalPath(path, streamProvider);
-        OrcReader reader = getReader(new File(path));
-        ColumnMetadata<OrcType> types = reader.getFooter().getTypes();
-        List<Type> readTypes = reader.getRootColumn().getNestedColumns().stream()
-                .map(orcColumn -> fromOrcType(types.get(orcColumn.getColumnId()), types))
-                .collect(Collectors.toList());
-
-        try {
-            OrcRecordReader recordReader = reader.createRecordReader(
-                    reader.getRootColumn().getNestedColumns(),
-                    readTypes,
-                    OrcPredicate.TRUE,
-                    UTC,
-                    newSimpleAggregatedMemoryContext(),
-                    INITIAL_BATCH_SIZE,
-                    OrcPlugin::handleException);
-            List<Page> result = new LinkedList<>();
-            Page page;
-            while ((page = recordReader.nextPage()) != null) {
-                result.add(page.getLoadedPage());
-            }
-            return result;
+        try (ClosableFile file = getLocalFile(path, streamProvider)) {
+            OrcReader reader = getReader(file.getFile());
+            ColumnMetadata<OrcType> types = reader.getFooter().getTypes();
+            return reader.getRootColumn().getNestedColumns().stream()
+                    .map(orcColumn -> new StorageColumn(
+                            orcColumn.getColumnName(),
+                            fromOrcType(types.get(orcColumn.getColumnId()), types)))
+                    .collect(Collectors.toList());
         }
         catch (IOException e) {
             throw new RuntimeException(e);
         }
     }
 
-    private String getLocalPath(String path, Function<String, InputStream> streamProvider)
+    @Override
+    public Iterable<Page> getPagesIterator(String path, Function<String, InputStream> streamProvider)
     {
-        if (path.startsWith("http://") || path.startsWith("https://") || path.startsWith("hdfs://") || path.startsWith("s3a://") || path.startsWith("s3://")) {
-            try (AutoDeletingTempFile tempFile = new AutoDeletingTempFile()) {
-                Files.copy(streamProvider.apply(path), tempFile.getFile().toPath(), StandardCopyOption.REPLACE_EXISTING);
-                return tempFile.getFile().getPath();
+        try (ClosableFile file = getLocalFile(path, streamProvider)) {
+            OrcReader reader = getReader(file.getFile());
+            ColumnMetadata<OrcType> types = reader.getFooter().getTypes();
+            List<Type> readTypes = reader.getRootColumn().getNestedColumns().stream()
+                    .map(orcColumn -> fromOrcType(types.get(orcColumn.getColumnId()), types))
+                    .collect(Collectors.toList());
+            try {
+                OrcRecordReader recordReader = reader.createRecordReader(
+                        reader.getRootColumn().getNestedColumns(),
+                        readTypes,
+                        OrcPredicate.TRUE,
+                        UTC,
+                        newSimpleAggregatedMemoryContext(),
+                        INITIAL_BATCH_SIZE,
+                        OrcPlugin::handleException);
+                List<Page> result = new ArrayList<>();
+                Page page;
+                while ((page = recordReader.nextPage()) != null) {
+                    result.add(page.getLoadedPage());
+                }
+                return result;
             }
             catch (IOException e) {
                 throw new RuntimeException(e);
             }
         }
-        if (path.startsWith("file:")) {
-            return path.substring(5);
+        catch (IOException e) {
+            throw new RuntimeException(e);
         }
-        return path;
+    }
+
+    private ClosableFile getLocalFile(String path, Function<String, InputStream> streamProvider)
+            throws IOException
+    {
+        if (path.startsWith("http://") || path.startsWith("https://") || path.startsWith("hdfs://") || path.startsWith("s3a://") || path.startsWith("s3://")) {
+            AutoDeletingTempFile tempFile = new AutoDeletingTempFile();
+            Files.copy(streamProvider.apply(path), tempFile.getFile().toPath(), StandardCopyOption.REPLACE_EXISTING);
+            return tempFile;
+        }
+        if (path.startsWith("file:")) {
+            return () -> new File(path.substring(5));
+        }
+        throw new IllegalArgumentException(format("Unsupported schema %s", path.split(":", 2)[0]));
     }
 
     private OrcReader getReader(File file)
@@ -136,8 +140,20 @@ public class OrcPlugin
         return new TrinoException(GENERIC_INTERNAL_ERROR, "Failed to read temporary data", e);
     }
 
+    public interface ClosableFile
+            extends AutoCloseable
+    {
+        File getFile();
+
+        @Override
+        default void close()
+                throws IOException
+        {
+        }
+    }
+
     public static class AutoDeletingTempFile
-            implements AutoCloseable
+            implements ClosableFile
     {
         private final File file;
 
@@ -147,6 +163,7 @@ public class OrcPlugin
             file = File.createTempFile("trino-storage-", ".orc");
         }
 
+        @Override
         public File getFile()
         {
             return file;

--- a/src/test/java/org/ebyhr/trino/storage/TestStorageConnector.java
+++ b/src/test/java/org/ebyhr/trino/storage/TestStorageConnector.java
@@ -70,6 +70,9 @@ public final class TestStorageConnector
         assertQuery(
                 format("SELECT * FROM storage.orc.\"%s\" WHERE x = 1658882660", toAbsolutePath("example-data/apache-lz4.orc")),
                 "VALUES (1658882660, 639, -5557347160648450358)");
+        assertQuery(
+                format("SELECT * FROM storage.orc.\"%s\" WHERE x = 1658882660", toRemotePath("example-data/apache-lz4.orc")),
+                "VALUES (1658882660, 639, -5557347160648450358)");
     }
 
     @Test
@@ -91,5 +94,10 @@ public final class TestStorageConnector
     private static String toAbsolutePath(String resourceName)
     {
         return Resources.getResource(resourceName).toString();
+    }
+
+    private static String toRemotePath(String resourceName)
+    {
+        return format("https://github.com/snowlift/trino-storage/raw/4c381eca1fa44b22372300659a937a57550c90b9/src/test/resources/%s", resourceName);
     }
 }

